src/meta-srv/src/handler/collect_region_leader.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use api::v1::meta::{HeartbeatRequest, Role};
+use common_meta::region_registry::LeaderRegion;
+use store_api::region_engine::RegionRole;
+
+use crate::error::Result;
+use crate::handler::{HandleControl, HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::Context;
+
+pub struct CollectLeaderRegionHandler;
+
+#[async_trait::async_trait]
+impl HeartbeatHandler for CollectLeaderRegionHandler {
+    fn is_acceptable(&self, role: Role) -> bool {
+        role == Role::Datanode
+    }
+
+    async fn handle(
+        &self,
+        _req: &HeartbeatRequest,
+        ctx: &mut Context,
+        acc: &mut HeartbeatAccumulator,
+    ) -> Result<HandleControl> {
+        let Some(current_stat) = acc.stat.as_ref() else {
+            return Ok(HandleControl::Continue);
+        };
+
+        let mut key_values = Vec::with_capacity(current_stat.region_stats.len());
+        for stat in current_stat.region_stats.iter() {
+            if stat.role != RegionRole::Leader {
+                continue;
+            }
+            let value = LeaderRegion {
+                datanode_id: current_stat.id,
+                manifest_version: stat.manifest_version,
+            };
+            key_values.push((stat.id, value));
+        }
+
+        ctx.leader_region_registry.batch_put(key_values);
+
+        Ok(HandleControl::Continue)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Do we need to add a test that the manifest_version in stat is less than the version in the registry?

Reply:
resolved in [0295477](https://github.com/GreptimeTeam/greptimedb/pull/5762/commits/029547737c67fb1a5e88abc5580eae21ad8694af)